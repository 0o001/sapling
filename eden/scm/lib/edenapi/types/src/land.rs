/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#[cfg(any(test, feature = "for-tests"))]
use quickcheck::Arbitrary;
use serde_derive::Deserialize;
use serde_derive::Serialize;
use std::collections::HashMap;
use type_macros::auto_wire;
use types::HgId;

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub struct PushVar {
    #[id(0)]
    pub key: String,

    #[id(1)]
    pub value: String,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for PushVar {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        PushVar {
            key: Arbitrary::arbitrary(g),
            value: Arbitrary::arbitrary(g),
        }
    }
}

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub struct LandStackRequest {
    #[id(0)]
    pub bookmark: String,

    #[id(1)]
    pub head: HgId,

    #[id(2)]
    pub base: HgId,

    #[id(4)]
    pub pushvars: Vec<PushVar>,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for LandStackRequest {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        LandStackRequest {
            bookmark: Arbitrary::arbitrary(g),
            head: Arbitrary::arbitrary(g),
            base: Arbitrary::arbitrary(g),
            pushvars: Arbitrary::arbitrary(g),
        }
    }
}

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub struct LandStackResponse {
    #[id(0)]
    pub new_head: HgId,

    #[id(1)]
    pub old_to_new_hgids: HashMap<HgId, HgId>,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for LandStackResponse {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        LandStackResponse {
            new_head: Arbitrary::arbitrary(g),
            old_to_new_hgids: Arbitrary::arbitrary(g),
        }
    }
}
