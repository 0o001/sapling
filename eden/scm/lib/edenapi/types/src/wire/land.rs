/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

pub use crate::land::WireLandStackRequest;
pub use crate::land::WireLandStackResponse;
pub use crate::land::WirePushVar;

#[cfg(test)]
mod tests {
    use super::*;
    use crate::wire::tests::auto_wire_tests;

    auto_wire_tests!(WirePushVar, WireLandStackRequest, WireLandStackResponse,);
}
