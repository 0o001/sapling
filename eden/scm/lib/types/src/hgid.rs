/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::{
    fmt::{self, Debug, Display},
    io::{self, Read, Write},
};

use failure::{Fail, Fallible as Result};
use serde_derive::{Deserialize, Serialize};

#[cfg(any(test, feature = "for-tests"))]
use rand::RngCore;

#[cfg(any(test, feature = "for-tests"))]
use std::collections::HashSet;

#[derive(Debug, Fail)]
#[fail(display = "HgId Error: {:?}", _0)]
struct HgIdError(String);

const HEX_CHARS: &[u8] = b"0123456789abcdef";

/// A 20-byte identifier, often a hash. Nodes are used to uniquely identify
/// commits, file versions, and many other things.
#[derive(
    Clone,
    Copy,
    Eq,
    Hash,
    Ord,
    PartialEq,
    PartialOrd,
    Serialize,
    Deserialize
)]
pub struct HgId([u8; HgId::len()]);

/// The nullid (0x00) is used throughout Mercurial to represent "None".
/// (For example, a commit will have a nullid p2, if it has no second parent).
pub const NULL_ID: HgId = HgId([0; HgId::len()]);

impl HgId {
    pub fn null_id() -> &'static Self {
        &NULL_ID
    }

    pub fn is_null(&self) -> bool {
        self == &NULL_ID
    }

    pub const fn len() -> usize {
        20
    }

    pub const fn hex_len() -> usize {
        40
    }

    pub fn from_slice(bytes: &[u8]) -> Result<Self> {
        if bytes.len() != HgId::len() {
            return Err(HgIdError(format!("invalid hgid length {:?}", bytes.len())).into());
        }

        let mut fixed_bytes = [0u8; HgId::len()];
        fixed_bytes.copy_from_slice(bytes);
        Ok(HgId(fixed_bytes))
    }

    pub fn from_byte_array(bytes: [u8; HgId::len()]) -> Self {
        HgId(bytes)
    }

    // Taken from Mononoke
    pub fn from_str(s: &str) -> Result<Self> {
        if s.len() != HgId::hex_len() {
            return Err(HgIdError(format!("invalid string length {:?}", s.len())).into());
        }

        let mut ret = HgId([0u8; HgId::len()]);

        for idx in 0..ret.0.len() {
            ret.0[idx] = match u8::from_str_radix(&s[(idx * 2)..(idx * 2 + 2)], 16) {
                Ok(v) => v,
                Err(_) => return Err(HgIdError(format!("bad digit")).into()),
            }
        }

        Ok(ret)
    }

    pub fn to_hex(&self) -> String {
        let mut v = Vec::with_capacity(HgId::hex_len());
        for &byte in self.as_ref() {
            v.push(HEX_CHARS[(byte >> 4) as usize]);
            v.push(HEX_CHARS[(byte & 0xf) as usize]);
        }

        unsafe { String::from_utf8_unchecked(v) }
    }

    #[cfg(any(test, feature = "for-tests"))]
    pub fn random(rng: &mut dyn RngCore) -> Self {
        let mut bytes = [0; HgId::len()];
        rng.fill_bytes(&mut bytes);
        loop {
            let hgid = HgId::from(&bytes);
            if !hgid.is_null() {
                return hgid;
            }
        }
    }

    #[cfg(any(test, feature = "for-tests"))]
    pub fn random_distinct(rng: &mut dyn RngCore, count: usize) -> Vec<Self> {
        let mut nodes = Vec::new();
        let mut nodeset = HashSet::new();
        while nodes.len() < count {
            let hgid = HgId::random(rng);
            if !nodeset.contains(&hgid) {
                nodeset.insert(hgid.clone());
                nodes.push(hgid);
            }
        }
        nodes
    }
}

impl Display for HgId {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        Display::fmt(&self.to_hex(), fmt)
    }
}

impl Debug for HgId {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "HgId({:?})", &self.to_hex())
    }
}

impl Default for HgId {
    fn default() -> HgId {
        NULL_ID
    }
}

impl<'a> From<&'a [u8; HgId::len()]> for HgId {
    fn from(bytes: &[u8; HgId::len()]) -> HgId {
        HgId(bytes.clone())
    }
}

impl AsRef<[u8]> for HgId {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}

pub trait WriteHgIdExt {
    /// Write a ``HgId`` directly to a stream.
    ///
    /// # Examples
    ///
    /// ```
    /// use types::hgid::{HgId, WriteHgIdExt};
    /// let mut v = vec![];
    ///
    /// let n = HgId::null_id();
    /// v.write_hgid(&n).expect("writing a hgid to a vec should work");
    ///
    /// assert_eq!(v, vec![0; HgId::len()]);
    /// ```
    fn write_hgid(&mut self, value: &HgId) -> io::Result<()>;
}

impl<W: Write + ?Sized> WriteHgIdExt for W {
    fn write_hgid(&mut self, value: &HgId) -> io::Result<()> {
        self.write_all(&value.0)
    }
}

pub trait ReadHgIdExt {
    /// Read a ``HgId`` directly from a stream.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::io::Cursor;
    /// use types::hgid::{HgId, ReadHgIdExt};
    /// let mut v = vec![0; HgId::len()];
    /// let mut c = Cursor::new(v);
    ///
    /// let n = c.read_hgid().expect("reading a hgid from a vec should work");
    ///
    /// assert_eq!(&n, HgId::null_id());
    /// ```
    fn read_hgid(&mut self) -> io::Result<HgId>;
}

impl<R: Read + ?Sized> ReadHgIdExt for R {
    fn read_hgid(&mut self) -> io::Result<HgId> {
        let mut hgid = HgId([0u8; HgId::len()]);
        self.read_exact(&mut hgid.0)?;
        Ok(hgid)
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl quickcheck::Arbitrary for HgId {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        let mut bytes = [0u8; HgId::len()];
        g.fill_bytes(&mut bytes);
        HgId::from(&bytes)
    }
}

#[cfg(any(test, feature = "for-tests"))]
pub mod mocks {
    use super::HgId;

    pub const ONES: HgId = HgId([0x11; HgId::len()]);
    pub const TWOS: HgId = HgId([0x22; HgId::len()]);
    pub const THREES: HgId = HgId([0x33; HgId::len()]);
    pub const FOURS: HgId = HgId([0x44; HgId::len()]);
    pub const FIVES: HgId = HgId([0x55; HgId::len()]);
    pub const SIXES: HgId = HgId([0x66; HgId::len()]);
    pub const SEVENS: HgId = HgId([0x77; HgId::len()]);
    pub const EIGHTS: HgId = HgId([0x88; HgId::len()]);
    pub const NINES: HgId = HgId([0x99; HgId::len()]);
    pub const AS: HgId = HgId([0xAA; HgId::len()]);
    pub const BS: HgId = HgId([0xAB; HgId::len()]);
    pub const CS: HgId = HgId([0xCC; HgId::len()]);
    pub const DS: HgId = HgId([0xDD; HgId::len()]);
    pub const ES: HgId = HgId([0xEE; HgId::len()]);
    pub const FS: HgId = HgId([0xFF; HgId::len()]);
}

#[cfg(test)]
mod tests {
    use super::*;

    use quickcheck::quickcheck;

    #[test]
    fn test_incorrect_length() {
        HgId::from_slice(&[0u8; 25]).expect_err("bad slice length");
    }

    quickcheck! {
        fn test_from_slice(hgid: HgId) -> bool {
            hgid == HgId::from_slice(hgid.as_ref()).expect("from_slice")
        }
    }
}
