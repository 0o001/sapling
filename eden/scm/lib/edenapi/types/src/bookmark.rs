/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#[cfg(any(test, feature = "for-tests"))]
use quickcheck::Arbitrary;
use serde_derive::Deserialize;
use serde_derive::Serialize;
use type_macros::auto_wire;
use types::hgid::HgId;

use crate::land::PushVar;

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub struct BookmarkRequest {
    #[id(0)]
    pub bookmarks: Vec<String>,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for BookmarkRequest {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        BookmarkRequest {
            bookmarks: Arbitrary::arbitrary(g),
        }
    }
}

#[auto_wire]
#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
#[derive(Serialize, Deserialize)]
pub struct BookmarkEntry {
    #[id(1)]
    pub bookmark: String,
    #[id(2)]
    pub hgid: Option<HgId>,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for BookmarkEntry {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        BookmarkEntry {
            bookmark: Arbitrary::arbitrary(g),
            hgid: Arbitrary::arbitrary(g),
        }
    }
}

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub struct SetBookmarkRequest {
    #[id(0)]
    pub bookmark: String,

    #[id(1)]
    pub to: Option<HgId>,

    #[id(2)]
    pub from: Option<HgId>,

    #[id(4)]
    pub pushvars: Vec<PushVar>,
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for SetBookmarkRequest {
    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
        SetBookmarkRequest {
            bookmark: Arbitrary::arbitrary(g),
            to: Arbitrary::arbitrary(g),
            from: Arbitrary::arbitrary(g),
            pushvars: Arbitrary::arbitrary(g),
        }
    }
}
