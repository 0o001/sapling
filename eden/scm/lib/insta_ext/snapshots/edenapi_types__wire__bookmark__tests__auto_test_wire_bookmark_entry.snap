---
source: bookmark.rs
expression: "WireBookmarkEntry::arbitrary(&mut g)"

---
{"1":"Â€","2":null}
